AnkiHubSoftware/ankihub_addon
# chore: Fix Anki 23.12 incompatibility

A warning dialog is shown when syncing with AnkiHub while using the 23.12 Beta 2 version of Anki:
<img src="https://github.com/ankipalace/ankihub_addon/assets/31575114/05813a1e-515c-49f0-862d-4cfdd97def0e" width="500" />

The reason is that Anki now shows this warning when `aqt.taskman.run_in_background()` is called in a background thread. The add-on currently does this when starting the media sync after the updates for AnkiHub decks were applied.

## Related issues
Community topic with an error report:
https://community.ankihub.net/t/anki-23-12-beta-2-error/122545


## Proposed changes
- Use `aqt.mw.taskman.run_on_main` so that `media_sync.start_media_download` is started on the main thread.
  Currently `_AnkiHubDeckUpdater.update_decks_and_media` is called in a background thread and it calls `media_sync.start_media_download()`. 
  Then `media_sync.start_media_download()` starts `media_sync._update_deck_media_and_download_missing_media` in a background thread (using `AddonQueryOp`).
  Anki doesn't allow background threads starting other background threads, so it shown the warning.


## How to reproduce
- Create a new virtual python environment
- Change the anki and aqt version to 23.12b2 in `requirements/base.txt`
- Install the requirements: `pip install -r requirements/dev.txt`
- Launch Anki with the add-on
- Sync with AnkiHub
- The warning dialog is shown

With the changes in this PR the warning dialog is not shown and the media sync is run.

## Comments
I reviewed the code base to see if we have the same problem in other places and as far as I can tell it only occurs in this one place.
